### PR TITLE
Account for constant equivalence properties in union, tests

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/12446
Closes https://github.com/apache/datafusion/pull/12414

## Rationale for this change

Fixes a bug: https://github.com/apache/datafusion/issues/12446

Note it doesn't include end to end tests as they hit a stack overflow (tracked by https://github.com/apache/datafusion/issues/12700)

I don't think we should merge this PR until  https://github.com/apache/datafusion/issues/12700 is fixed as that would turn what is now a plan runtime error into a stack overflow for certain expressions, which seems like a regression to me

## What changes are included in this PR?

1. Update the code that calculates the output `EquivalenceProperties` for a union to account for constants and "augment" the sort orders
2. Add additional tests

## Are these changes tested?
Yes

## Are there any user-facing changes?

Fixed bug

### CODE REVIEW

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 1]
```diff
@@ -2617,379 +2633,356 @@ mod tests {
...
-        let schema = create_test_schema()?;
+    #[test]
+    fn test_union_equivalence_properties_multi_children_1() {
+        let schema = create_test_schema().unwrap();
         let schema2 = append_fields(&schema, "1");
         let schema3 = append_fields(&schema, "2");
-        let test_cases = vec![
-            // --------- TEST CASE 1 ----------
-            (
+        UnionEquivalenceTest::new(&schema)
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:2641]: I wanted more flexibility with the tests, and I am glad I did this as I hadn't seen the existing tests for constants

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 2]
```diff
@@ -1611,58 +1611,62 @@ impl Hash for ExprWrapper {
...
-            ordering.pop();
-        }
-        // There is a non-trivial satisfied prefix, add it as a valid ordering:
-        if !ordering.is_empty() {
-            orderings.push(ordering);
-        }
-    }
-    let mut eq_properties = EquivalenceProperties::new(lhs.schema);
-    eq_properties.constants = constants;
+    let mut orderings = UnionEquivalentOrderingBuilder::new();
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:1654]: this is the key change that encapsulates the sort expression calculation

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 3]
```diff
@@ -2877,21 +3078,18 @@ mod tests {
...
+                // corresponding schemas.
                 vec![vec!["a"]],
                 vec![],
             )
             .run()
     }
 
     #[test]
-    #[ignore]
-    // ignored due to https://github.com/apache/datafusion/issues/12446
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:2893]: the tests now pass 🎉  -- I also renamed them to try and be more specific about what they are testing

File: datafusion/physical-expr/src/equivalence/class.rs [hunk 4]
```diff
@@ -101,6 +118,31 @@ impl ConstExpr {
...
         })
     }
+
+    /// Returns true if this constant expression is equal to the given expression
+    pub fn eq_expr(&self, other: impl AsRef<dyn PhysicalExpr>) -> bool {
+        self.expr.eq(other.as_ref().as_any())
+    }
+
+    /// Returns a [`Display`]able list of `ConstExpr`.
+    pub fn format_list(input: &[ConstExpr]) -> impl Display + '_ {
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/class.rs:128]: this is just moved into a function from EquivalenceProperties

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 5]
```diff
@@ -3088,18 +3352,32 @@ mod tests {
...
-                    .expect("failed to calculate union equivalence properties");
-            assert_eq_properties_same(
-                &actual_properties,
-                &expected_properties,
-                format!(
-                    "expected: {expected_properties:?}\nactual:  {actual_properties:?}"
-                ),
-            );
+
+            // try all permutations of the children
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:3359]: I also updated the test runner to test with all child permutations (aka different orders) as the algorithm is not obviously symmetric

File: datafusion/physical-expr/src/equivalence/properties.rs [hunk 6]
```diff
@@ -1675,8 +1679,7 @@ pub fn calculate_union(
     eqps: Vec<EquivalenceProperties>,
     schema: SchemaRef,
 ) -> Result<EquivalenceProperties> {
-    // TODO: In some cases, we should be able to preserve some equivalence
-    //       classes. Add support for such cases.
+    // TODO avoid this clone by using `reduce`
```
Comment (alamb) [inline datafusion/physical-expr/src/equivalence/properties.rs:1682]: in https://github.com/apache/datafusion/pull/12722

**Review by berkaysynnada — APPROVED** [review]
Thanks @alamb for this great algorithm. I have tested a few tricky cases, and all gave the expected results. The new tests have also improved the coverage. Maybe @akurmustafa could check if we missed anything, but I believe it is solid and aligns with what I had in mind.

**Comment by alamb** [comment]
Sounds good -- thank you for the review @berkaysynnada 

I am currently working on https://github.com/apache/datafusion/issues/12700 which I think should be fixed too prior to merging this PR.

**Comment by alamb** [comment]
Now that https://github.com/apache/datafusion/issues/12700 has been fixed, I have merged this PR up from main and plan to merge it later today unless anyone would like additional time to comment

I am also happy to make changes as follow on PRs if needed

**Comment by alamb** [comment]
Thanks again @berkaysynnada

**Pull request merged**